### PR TITLE
Implement thermoplast (new version)

### PR DESCRIPTION
**Pull request by Oliveriver**
**Brief Description of What This PR Does**

Adds thermoplasts and thermosynthase as organelles which create ATP at a rate which scales with heat in a patch. See: https://forum.revolutionarygamesstudio.com/t/thermoplasts/898

Supercedes the old thermoplast PR: #1954

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

**Comment by Oliveriver** [comment]
Surprisingly, I was able to add all this quite easily. The models and GUI elements came from the old PR.

There are still a few issues, which is why I've marked this as a draft for now.

First, the thermosynthase text is cut off and the icon looks odd.


Second, I've no idea how sensible the Auto-Evo changes are. Scaling temperature efficiency to make AI cells develop thermosynthase at an acceptable rate proved quite hard. I'll take another look later.

Other than those and adjusting some code styling, I don't think there's anything else to do code-wise for this very simple thermosynthesis implementation.

**Comment by Kasterisk** [comment]
The game crashes whenever I place thermosynthase or thermoplast, the problem seems to be that the tooltips for both parts for the ATP balance bar is missing.

> First, the thermosynthase text is cut off and the icon looks oddly grungy.

Text getting cut off like that is unavoidable. Relevant issue: https://github.com/Revolutionary-Games/Thrive/issues/2192.

You need to turn on mipmaps for the thermosynthase icon to make it not look jaggy.

**Comment by Oliveriver** [comment]
> The game crashes whenever I place thermosynthase or thermoplast, the problem seems to be that the tooltips for both parts for the ATP balance bar is missing.

I had fixed that but forgot to commit the file. Should be all good now.

**Comment by hhyyrylainen** [comment]
> It seems Auto-Evo rarely produces species with thermosynthesis on its own, but if you add thermosynthetic organelles to your species, your population can get quite high.

Does it ever do that? If it doesn't then probably adjusting the random chance for thermoplast wasn't done, it was set to 0 when it wasn't implemented.

**Comment by Oliveriver** [comment]
> Does it ever do that? If it doesn't then probably adjusting the random chance for thermoplast wasn't done, it was set to 0 when it wasn't implemented.

It does sometimes, though not often. And ChanceToCreate is set properly for the thermoplast.

File: simulation_parameters/microbe_stage/organelles.json [hunk 1]
```diff
@@ -308,6 +308,41 @@
...
+        "processes": {
+            "bacterial_thermosynthesis": 1
+        },
+        "components": {
+            "storage": {
+                "capacity": 0.5
+            },
+            "processor": {
+                "colourChangeFactor": 1
+            }
```
Comment (hhyyrylainen) [inline simulation_parameters/microbe_stage/organelles.json:332]: Does some other organelle specify this? As far as I remember and just quickly checked, this kind of component doesn't exist, so I wouldn't specify it here for this new organelle.

File: simulation_parameters/microbe_stage/organelles.json [hunk 2]
```diff
@@ -732,27 +767,43 @@
...
-        "processes": {},
-        "components": {},
+        "processes": {
+            "thermosynthesis": 1
+        },
+        "components": {
+            "storage": {
+                "capacity": 1
+            },
+            "processor": {
```
Comment (hhyyrylainen) [inline simulation_parameters/microbe_stage/organelles.json:792]: Same here.

File: src/gui_common/CompoundAmount.cs [hunk 3]
```diff
@@ -166,7 +166,11 @@ private void UpdateLabel()
         }
 
         string numberPart;
-        if (UsePercentageDisplay)
+        if (!string.IsNullOrEmpty(compound!.Unit))
+        {
+            numberPart = Math.Round(amount) + " " + compound.Unit;
```
Comment (hhyyrylainen) [inline src/gui_common/CompoundAmount.cs:171]: I think this should use the translation system, a key like "VALUE_WITH_UNIT", which would be `{0} {1}` for English, should probably be used as some language might want to switch the order of those here.

File: src/microbe_stage/Compound.cs [hunk 4]
```diff
@@ -30,6 +30,11 @@ public class Compound : IRegistryType
 
     public bool IsEnvironmental;
 
+    /// <summary>
+    ///   Unit for this compound, if applicable (e.g. °C for temperature)
+    /// </summary>
+    public string Unit = null!;
```
Comment (hhyyrylainen) [inline src/microbe_stage/Compound.cs:36]:
```suggestion
    public string? Unit;
```
As this can be often null, marking that there is no unit.

File: src/microbe_stage/MicrobeHUD.cs [hunk 5]
```diff
@@ -246,7 +247,7 @@ public class MicrobeHUD : Control
     private ProgressBar oxygenBar = null!;
     private ProgressBar co2Bar = null!;
     private ProgressBar nitrogenBar = null!;
-    private ProgressBar temperature = null!;
+    private ProgressBar temperatureLabel = null!;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:250]: This is of type ProgressBar, why did you add the Label suffix? That should only be used for things that are actual labels...

File: src/microbe_stage/MicrobeHUD.cs [hunk 6]
```diff
@@ -733,7 +735,8 @@ public void UpdateEnvironmentalBars(BiomeConditions biome)
 
         sunlightLabel.GetNode<Label>("Value").Text =
             string.Format(CultureInfo.CurrentCulture, percentageFormat, sunlightPercentage);
-        temperature.GetNode<Label>("Value").Text = averageTemperature + " °C";
+        temperatureLabel.GetNode<Label>("Value").Text =
+            averageTemperature + " " + temperature.Unit;
```
Comment (hhyyrylainen) [inline src/microbe_stage/MicrobeHUD.cs:739]: Here should use the translation as well. Actually a few lines above this, the percentageFormat use shows well how the unit format should also be made to work.

File: src/microbe_stage/ProcessSystem.cs [hunk 7]
```diff
@@ -222,9 +223,12 @@ void MakeSureResultExists(Compound compound)
 
             // Environmental compound that can limit the rate
 
-            var availableInEnvironment = GetDissolvedInBiome(input.Key, biome);
+            var availableInEnvironment = GetAmbientInBiome(input.Key, biome);
 
-            var availableRate = availableInEnvironment / input.Value;
+            float availableRate;
+            availableRate = input.Key == Temperature ?
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:229]:
```suggestion
            var availableRate = input.Key == Temperature ?
```

File: src/microbe_stage/editor/PatchMapEditorComponent.cs [hunk 8]
```diff
@@ -395,7 +395,8 @@ private void UpdatePatchDetails(Patch patch)
         var percentageFormat = TranslationServer.Translate("PERCENTAGE_VALUE");
 
         // Atmospheric gasses
-        patchTemperature.Text = patch.Biome.AverageTemperature + " °C";
+        var temperature = SimulationParameters.Instance.GetCompound("temperature");
+        patchTemperature.Text = patch.Biome.Compounds[temperature].Ambient + temperature.Unit;
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/PatchMapEditorComponent.cs:399]: Format here as well.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Overall good work, I just have a few small comments. Also I'll push a commit where I fix spelling warnings by adding some stuff to the dictionary and I also did some Godot's automatically wanted changes to the import files.

File: simulation_parameters/microbe_stage/organelles.json (hunk 1 above)
Comment (Oliveriver) [reply simulation_parameters/microbe_stage/organelles.json:332]: Almost all of them do. You're right that it doesn't look like it's actually used, but that sounds like a problem for another PR.

Comment (hhyyrylainen) [reply simulation_parameters/microbe_stage/organelles.json:332]: Fine, https://github.com/Revolutionary-Games/Thrive/issues/3476

**Review by hhyyrylainen — APPROVED** [review]
I think all of my code concerns were addressed. I did also quickly playtest the prokaryotic version and it seemed to work, but I'll wait until tomorrow morning before merging in case someone still wants to take a look at this.

Oh and, this has conflicts in the translation files.

**Comment by buckly90** [comment]
Great work on this. Folks will be ecstatic to have a new source of energy to play with.

You did a great job on balancing the parts, so there was hardly any need to change anything. However, the mass for the parts were quite high so I changed them to reflect the mass of photosynthetic parts for the time being.

Thank you
